File: Day1_PokemonApp/PokemonApp/Data Models/PaginatedResponse.swift
//
//  PaginatedResponse.swift
//  PokemonApp
//
//  Created by Sanket Sonje on 26/01/25.
//

/// A generic structure that represents a paginated response from the PokeAPI.
/// This structure is used to decode paginated API responses where the results
/// can be of any decodable type.
///
/// Example usage:
/// ```
/// let response: PaginatedResponse<PokemonListItem> = try decoder.decode(...)
/// ```
struct PaginatedResponse<T: Decodable>: Decodable {

    /// The total count of items available in the API
    let count: Int

    /// URL string for the next page of results
    /// Will be nil if there are no more pages after this one
    let next: String?

    /// URL string for the previous page of results
    /// Will be nil if this is the first page
    let previous: String?

    /// Array of decoded items of type T
    /// Contains the actual data for the current page
    let results: [T]
}

File: Android_PokemonApp/app/src/main/java/com/example/android_pokemonapp/data/model/PaginatedResponse.kt
package com.example.android_pokemonapp.data.model

import com.google.gson.annotations.SerializedName

/**
 * Represents a paginated response from the PokemonAPI.
 *
 * Handles the pagination of Pokemon list data, providing navigation links and
 * metadata for traversing through the complete Pokemon collection.
 *
 * @property count Total number of available Pokemon in the API
 * @property next URL for the next page of results, null if this is the last page
 * @property previous URL for the previous page of results, null if this is the first page
 * @property results List of Pokemon items for the current page
 * @see PokemonListItem
 */
data class PaginatedResponse(
	@SerializedName("count")
	val count: Int,

	@SerializedName("next")
	val next: String?,

	@SerializedName("previous")
	val previous: String?,

	@SerializedName("results")
	val results: List<PokemonListItem>
)
